C error - fixed arrays in if/match expressions

### Describe the bug

The compiler outputs a C error.

### Reproduction Steps

Command `v file.v`

```
_ := if u64(0) == u64(0) {
        [0, 1]!
} else {
        return
}

_ := match u64(0) {
         u64(0) {
                [0, 1]!
        } else {
                return
        }
}
```

### Expected Behavior

No C error.

### Current Behavior

```
❯ v  test.v
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/test.01JJF6R998VK7HE2AQFFW89MS0.tmp.c:5383: error: lvalue expected
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error:
==================
C error found. It should never happen, when compiling pure V code.
...
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.9 6ac1d08.d710d9e

### Environment details (OS name and version, etc.)

|V full version      |V 0.4.9 6ac1d08.d710d9e
|:-------------------|:-------------------
|OS                  |linux, "Void Linux"
|Processor           |4 cpus, 64bit, little endian, Intel(R) Pentium(R) Gold 7505 @ 2.00GHz
|Memory              |17.08GB/23.16GB
|                    |
|V executable        |/home/nopana/v/v
|V last modified time|2025-01-25 16:49:25
|                    |
|V home dir          |OK, value: /home/nopana/v
|VMODULES            |OK, value: /home/nopana/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|Current working dir |OK, value: /home/nopana/projects/vreduce
|                    |
|Git version         |git version 2.47.1
|V git status        |weekly.2025.03-37-gd710d9ef
|.git/config present |true
|                    |
|cc version          |cc (GCC) 13.2.0
|gcc version         |gcc (GCC) 13.2.0
|clang version       |N/A
|tcc version         |tcc version 0.9.28rc 2024-07-31 HEAD@1cee0908 (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 0134e9b9
|emcc version        |N/A
|glibc version       |ldd (GNU libc) 2.39
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

The offending lines:
```c
	Array_fixed_int_2 _t1; /* if prepend */
	if (((u64)(0U)) == ((u64)(0U))) {
		_t1 = {0, 1};  // tcc: lvalue expected
	//      ^ gcc: expected expression before '{' token; clang: expected expression
	} else {
		return;
	}
	{Array_fixed_int_2 _ = _t1;}
	//                     ^ gcc: invalid initializer; clang: array initializer must be an initializer list or wide string literal
	;
	u64 _t2 = ((u64)(0U));
	Array_fixed_int_2 _t3 = {0};
	
	if (_t2 == (((u64)(0U)))) {
		_t3 = {0, 1};
	//      ^ gcc: expected expression before '{' token; clang: expected expression
	}
	else {
		return;
	}{Array_fixed_int_2 _ = _t3;}
	//                     ^ gcc: invalid initializer; clang: array initializer must be an initializer list or wide string literal
	;
```

How `_t1` isn't an lvalue, I have no idea.

<details>
<summary>Raw error messages from invoking the compilers directly</summary>
tcc:
```
error.c:5284: error: lvalue expected
```

gcc:
```
error.c: In function ‘main__main’:
error.c:5284:23: error: expected expression before ‘{’ token
 5284 |                 _t1 = {0, 1};
      |                       ^
error.c:5288:32: error: invalid initializer
 5288 |         {Array_fixed_int_2 _ = _t1;}
      |                                ^~~
error.c:5294:23: error: expected expression before ‘{’ token
 5294 |                 _t3 = {0, 1};
      |                       ^
error.c:5298:33: error: invalid initializer
 5298 |         }{Array_fixed_int_2 _ = _t3;}
      |                                 ^~~
```

clang:
```
error.c:5284:9: error: expected expression
 5284 |                 _t1 = {0, 1};
      |                       ^
error.c:5288:21: error: array initializer must be an initializer list or wide string literal
 5288 |         {Array_fixed_int_2 _ = _t1;}
      |                            ^
error.c:5294:9: error: expected expression
 5294 |                 _t3 = {0, 1};
      |                       ^
error.c:5298:22: error: array initializer must be an initializer list or wide string literal
 5298 |         }{Array_fixed_int_2 _ = _t3;}
      |                             ^
4 errors generated.
```

error.c was generated with `v -o error.c error.v`.
error.v is a direct copy-paste from the example in the issue.
</details>

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -526,7 +526,13 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 				} else {
 					g.write('{${styp} _ = ')
 				}
-				g.expr(val)
+				if unaliased_right_sym.info is ast.ArrayFixed {
+					tmp_var := g.expr_with_var(val, var_type, false)
+					g.fixed_array_var_init(tmp_var, false, unaliased_right_sym.info.elem_type,
+						unaliased_right_sym.info.size)
+				} else {
+					g.expr(val)
+				}
 				g.writeln(';}')
 			}
 		} else if node.op == .assign && !g.pref.translated && (is_fixed_array_init

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -526,8 +526,7 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 				} else {
 					g.write('{${styp} _ = ')
 				}
-				if val !in [ast.StructInit, ast.ArrayInit]
-					&& unaliased_right_sym.info is ast.ArrayFixed {
+				if val in [ast.MatchExpr, ast.IfExpr] && unaliased_right_sym.info is ast.ArrayFixed {
 					tmp_var := g.expr_with_var(val, var_type, false)
 					g.fixed_array_var_init(tmp_var, false, unaliased_right_sym.info.elem_type,
 						unaliased_right_sym.info.size)
@@ -918,6 +917,11 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 							g.array_init(val, c_name(ident.name))
 						} else if val_type.has_flag(.shared_f) {
 							g.expr_with_cast(val, val_type, var_type)
+						} else if val in [ast.MatchExpr, ast.IfExpr]
+							&& unaliased_right_sym.info is ast.ArrayFixed {
+							tmp_var := g.expr_with_var(val, var_type, false)
+							g.fixed_array_var_init(tmp_var, false, unaliased_right_sym.info.elem_type,
+								unaliased_right_sym.info.size)
 						} else {
 							g.expr(val)
 						}

--- a/vlib/v/gen/c/assign.v
+++ b/vlib/v/gen/c/assign.v
@@ -526,7 +526,8 @@ fn (mut g Gen) assign_stmt(node_ ast.AssignStmt) {
 				} else {
 					g.write('{${styp} _ = ')
 				}
-				if unaliased_right_sym.info is ast.ArrayFixed {
+				if val !in [ast.StructInit, ast.ArrayInit]
+					&& unaliased_right_sym.info is ast.ArrayFixed {
 					tmp_var := g.expr_with_var(val, var_type, false)
 					g.fixed_array_var_init(tmp_var, false, unaliased_right_sym.info.elem_type,
 						unaliased_right_sym.info.size)

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -2114,6 +2114,9 @@ fn (mut g Gen) stmts_with_tmp_var(stmts []ast.Stmt, tmp_var string) bool {
 					g.stmt(stmt)
 				}
 			} else {
+				mut is_array_fixed_init := false
+				mut ret_type := ast.void_type
+
 				g.set_current_pos_as_last_stmt_pos()
 				g.skip_stmt_pos = true
 				mut is_noreturn := false
@@ -2121,11 +2124,23 @@ fn (mut g Gen) stmts_with_tmp_var(stmts []ast.Stmt, tmp_var string) bool {
 					is_noreturn = true
 				} else if stmt is ast.ExprStmt {
 					is_noreturn = is_noreturn_callexpr(stmt.expr)
+					if stmt.expr is ast.ArrayInit && stmt.expr.is_fixed {
+						is_array_fixed_init = true
+						ret_type = stmt.expr.typ
+					}
 				}
 				if !is_noreturn {
-					g.write('${tmp_var} = ')
+					if is_array_fixed_init {
+						g.write('memcpy(${tmp_var}, (${g.styp(ret_type)})')
+					} else {
+						g.write('${tmp_var} = ')
+					}
 				}
 				g.stmt(stmt)
+				if is_array_fixed_init {
+					lines := g.go_before_last_stmt().trim_right('; \n')
+					g.writeln('${lines}, sizeof(${tmp_var}));')
+				}
 				if !g.out.last_n(2).contains(';') {
 					g.writeln(';')
 				}

--- a/vlib/v/gen/c/if.v
+++ b/vlib/v/gen/c/if.v
@@ -20,6 +20,9 @@ fn (mut g Gen) need_tmp_var_in_if(node ast.IfExpr) bool {
 			if branch.stmts.len == 1 {
 				if branch.stmts[0] is ast.ExprStmt {
 					stmt := branch.stmts[0] as ast.ExprStmt
+					if stmt.expr is ast.ArrayInit && stmt.expr.is_fixed {
+						return true
+					}
 					if g.need_tmp_var_in_expr(stmt.expr) {
 						return true
 					}

--- a/vlib/v/gen/c/match.v
+++ b/vlib/v/gen/c/match.v
@@ -25,6 +25,9 @@ fn (mut g Gen) need_tmp_var_in_match(node ast.MatchExpr) bool {
 			if branch.stmts.len == 1 {
 				if branch.stmts[0] is ast.ExprStmt {
 					stmt := branch.stmts[0] as ast.ExprStmt
+					if stmt.expr is ast.ArrayInit && stmt.expr.is_fixed {
+						return true
+					}
 					if g.need_tmp_var_in_expr(stmt.expr) {
 						return true
 					}

--- a/vlib/v/tests/array_fixed_ternary_test.v
+++ b/vlib/v/tests/array_fixed_ternary_test.v
@@ -4,8 +4,8 @@ fn test_main() {
 
 	a := if true { [0]! } else { [1]! }
 	b := if true { [0] } else { [1] }
-	dump(a)
-	dump(b)
+	assert a.str() == '[0]'
+	assert b.str() == '[0]'
 }
 
 fn test_match() {
@@ -24,5 +24,5 @@ fn test_match() {
 		5 { ['Big', '5x5']! }
 		else { ['Large', '6x6']! }
 	}
-	dump(w)
+	assert w.str() == "['Large', '6x6']"
 }

--- a/vlib/v/tests/array_fixed_ternary_test.v
+++ b/vlib/v/tests/array_fixed_ternary_test.v
@@ -1,4 +1,28 @@
 fn test_main() {
 	_ := if true { [0]! } else { [1]! }
 	_ := if true { [0] } else { [1] }
+
+	a := if true { [0]! } else { [1]! }
+	b := if true { [0] } else { [1] }
+	dump(a)
+	dump(b)
+}
+
+fn test_match() {
+	grid_size := 1
+
+	_ := match grid_size {
+		3 { ['Small', '3x3']! }
+		4 { ['Classic', '4x4']! }
+		5 { ['Big', '5x5']! }
+		else { ['Large', '6x6']! }
+	}
+
+	w := match grid_size {
+		3 { ['Small', '3x3']! }
+		4 { ['Classic', '4x4']! }
+		5 { ['Big', '5x5']! }
+		else { ['Large', '6x6']! }
+	}
+	dump(w)
 }

--- a/vlib/v/tests/array_fixed_ternary_test.v
+++ b/vlib/v/tests/array_fixed_ternary_test.v
@@ -1,0 +1,4 @@
+fn test_main() {
+	_ := if true { [0]! } else { [1]! }
+	_ := if true { [0] } else { [1] }
+}

